packages/core/crates/core-path/src/channel_graph.rs

```diff
@@ -0,0 +1,144 @@
+    Status { old: ChannelStatus, new: ChannelStatus },
+
+    /// Channel balance has changed
+    CurrentBalance { old: Balance, new: Balance },
+
+    /// Channel epoch has changed
+    Epoch { old: u32, new: u32 },
+}
+
+impl Display for ChannelChange {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Status { old, new } => {
+                write!(f, "Status: {old} -> {new}")
+            }
+
+            CurrentBalance { old, new } => {
+                write!(f, "Balance: {old} -> {new}")
+            }
+
+            Epoch { old, new } => {
+                write!(f, "Epoch: {old} -> {new}")
+            }
+        }
+    }
+}
+
+impl ChannelGraph {
+    /// Maximum number of intermediate hops the automatic path finding algorithm will look for.
+    pub const INTERMEDIATE_HOPS: usize = 3;
+
+    /// Creates a new instance with the given self `Address`.
+    pub fn new(me: Address) -> Self {
+        Self {
+            me,
+            graph: DiGraphMap::default(),
+        }
+    }
+
+    /// Checks if the channel is incoming to or outgoing from this node
+    pub fn is_own_channel(&self, channel: &ChannelEntry) -> bool {
+        channel.destination == self.me || channel.source == self.me
+    }
+
+    /// Convenience method to get this node's own address
+    pub fn my_address(&self) -> Address {
+        self.me
+    }
+
+    /// Inserts or updates the given channel in the channel graph.
+    /// Returns a set of changes if the channel was already present in the graphs or
+    /// None if the channel was not previously present in the channel graph.
+    pub fn update_channel(&mut self, channel: ChannelEntry) -> Option<Vec<ChannelChange>> {
+        let weighted = WeightedChannel {
+            channel,
+            weight: 1_f64, // TODO: compute weight properly
+        };
+
+        if let Some(old_w_value) = self.graph.add_edge(channel.source, channel.destination, weighted) {
+            let mut ret = Vec::new();
+            let old_channel = old_w_value.channel;
+
+            if old_channel.status != channel.status {
+                ret.push(Status {
+                    old: old_channel.status,
+                    new: channel.status,
+                });
+            }
+
+            if old_channel.balance != channel.balance {
+                ret.push(CurrentBalance {
+                    old: old_channel.balance,
+                    new: channel.balance,
+                });
+            }
+
+            if old_channel.channel_epoch != channel.channel_epoch {
+                ret.push(Epoch {
+                    old: old_channel.channel_epoch.as_u32(),
+                    new: channel.channel_epoch.as_u32(),
+                });
+            }
+            debug!("channel update changes: {:?}", ret);
+
+            info!("updated {channel}: {} changes", ret.len());
+            Some(ret)
+        } else {
+            info!("new {channel}");
+            None
```

Comment:
Why prefer `None` over an empty `Vec`?

Reply:
Empty `Vec` means channel exists, but no changes were found. `None` means the channel did not exist before.